server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -242,38 +241,23 @@ public ShardIterator activeInitializingShardsRankedIt(
         return new PlainShardIterator(shardId, ordered);
     }
 
-    private static Set<String> getAllNodeIds(final List<ShardRouting> shards) {
-        final Set<String> nodeIds = new HashSet<>();
-        for (ShardRouting shard : shards) {
-            nodeIds.add(shard.currentNodeId());
-        }
-        return nodeIds;
-    }
-
-    private static Map<String, Optional<ResponseCollectorService.ComputedNodeStats>> getNodeStats(
-        final Set<String> nodeIds,
+    private static Map<String, ResponseCollectorService.ComputedNodeStats> getNodeStats(
+        final List<ShardRouting> shards,
         final ResponseCollectorService collector
     ) {
-
-        final Map<String, Optional<ResponseCollectorService.ComputedNodeStats>> nodeStats = Maps.newMapWithExpectedSize(nodeIds.size());
-        for (String nodeId : nodeIds) {
-            nodeStats.put(nodeId, collector.getNodeStatistics(nodeId));
+        final Map<String, ResponseCollectorService.ComputedNodeStats> nodeStats = new HashMap<>();
+        for (ShardRouting shard : shards) {
+            nodeStats.computeIfAbsent(shard.currentNodeId(), collector::getNodeStatistics);
         }
         return nodeStats;
     }
 
     private static Map<String, Double> rankNodes(
-        final Map<String, Optional<ResponseCollectorService.ComputedNodeStats>> nodeStats,
+        final Map<String, ResponseCollectorService.ComputedNodeStats> nodeStats,
         final Map<String, Long> nodeSearchCounts
     ) {
         final Map<String, Double> nodeRanks = Maps.newMapWithExpectedSize(nodeStats.size());
-        for (Map.Entry<String, Optional<ResponseCollectorService.ComputedNodeStats>> entry : nodeStats.entrySet()) {
-            Optional<ResponseCollectorService.ComputedNodeStats> maybeStats = entry.getValue();
-            maybeStats.ifPresent(stats -> {
-                final String nodeId = entry.getKey();
-                nodeRanks.put(nodeId, stats.rank(nodeSearchCounts.getOrDefault(nodeId, 0L)));
-            });
-        }
+        nodeStats.forEach((nodeId, value) -> nodeRanks.put(nodeId, value.rank(nodeSearchCounts.getOrDefault(nodeId, 0L))));
```

Comment:
Use `Maps.transformValues`?

Reply:
Idk, that one (if it doesn't inline) is kinda slow, particularly in the scenario we have here where we have tiny maps and frequently call the method. Not sure it's worth using an abstraction like that in somewhat critical code.

Reply:
Jup it's the method call to that function and in this specific example also just the method call to the utility on `Maps` in some cases because the map is generally tiny but this gets called a bunch of times.